### PR TITLE
Add template for accepting vector diffusivity in diffusive flux postprocessor

### PR DESCRIPTION
**Pull request by GiudGiud**
I only did the SideDiffusiveFlux one, and only RealVectorValue. 

Once this is validated, we can consider adding the tensor already and the interface version of this. 

I kind of wish I could have defined a partial specialization of the product member function, but I could not figure it out. Someone better than me at template meta-programming could let me know how to do that cleanly (eg not partially specializing the whole class first)

### CODE REVIEW

File: framework/src/postprocessors/SideDiffusiveFluxIntegral.C [hunk 1]
```diff
@@ -60,11 +62,36 @@ SideDiffusiveFluxIntegralTempl<is_ad>::computeQpIntegral()
...
     // FIXME Get the diffusion coefficient on the face, see #16809
-    return -MetaPhysicL::raw_value(_diffusion_coef[_qp]) * grad_u * _normals[_qp];
+    return -diffusivity_gradient_product(grad_u, MetaPhysicL::raw_value(_diffusion_coef[_qp])) *
+           _normals[_qp];
   }
   else
-    return -MetaPhysicL::raw_value(_diffusion_coef[_qp]) * _grad_u[_qp] * _normals[_qp];
+    return -diffusivity_gradient_product(_grad_u[_qp],
+                                         MetaPhysicL::raw_value(_diffusion_coef[_qp])) *
+           _normals[_qp];
```
Comment (GiudGiud) [inline framework/src/postprocessors/SideDiffusiveFluxIntegral.C:71]: I've tried changing this a couple times, clang won't let me

**Comment by GiudGiud** [comment]
Internal app test failures are unrelated, though that test does use this PP.

File: framework/src/postprocessors/SideDiffusiveFluxIntegral.C (hunk 1 above)
Comment (lindsayad) [reply framework/src/postprocessors/SideDiffusiveFluxIntegral.C:71]: What is the error?

File: framework/include/postprocessors/SideDiffusiveFluxIntegral.h [hunk 2]
```diff
@@ -36,5 +38,12 @@ class SideDiffusiveFluxIntegralTempl : public SideIntegralVariablePostprocessor
...
   MaterialPropertyName _diffusivity;
-  const GenericMaterialProperty<Real, is_ad> & _diffusion_coef;
+  const GenericMaterialProperty<T, is_ad> & _diffusion_coef;
+
+private:
+  /// Routine to get the diffusive flux with a Real diffusivity
+  RealVectorValue diffusivity_gradient_product(RealVectorValue grad_u, Real diffusivity);
+
+  /// Routine to get the diffusive flux with a RealVectorValue diffusivity
+  RealVectorValue diffusivity_gradient_product(RealVectorValue grad_u, RealVectorValue diffusivity);
```
Comment (lindsayad) [inline framework/include/postprocessors/SideDiffusiveFluxIntegral.h:48]: MOOSE style is camel case, so `diffusivityGradientProduct`

File: framework/src/postprocessors/SideDiffusiveFluxIntegral.C [hunk 3]
```diff
@@ -60,11 +62,36 @@ SideDiffusiveFluxIntegralTempl<is_ad>::computeQpIntegral()
...
+}
+
+template <bool is_ad, typename T>
+RealVectorValue
+SideDiffusiveFluxIntegralTempl<is_ad, T>::diffusivity_gradient_product(RealVectorValue grad_u,
+                                                                       RealVectorValue diffusivity)
+{
+  RealVectorValue d_grad_u = grad_u;
+  for (unsigned int i = 0; i < LIBMESH_DIM; i++)
+    d_grad_u *= diffusivity(i);
```
Comment (lindsayad) [inline framework/src/postprocessors/SideDiffusiveFluxIntegral.C:89]: This is not a component-wise multiplication. Isn't that what you want? What you are doing now will yield this vector:

(grad_u0 * d0 * d1 * d2, grad_u1 * d0 * d1 * d2, grad_u2 * d0 * d1 * d2)

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/src/postprocessors/SideDiffusiveFluxIntegral.C (hunk 1 above)
Comment (GiudGiud) [reply framework/src/postprocessors/SideDiffusiveFluxIntegral.C:71]: precheck whining

File: framework/src/postprocessors/SideDiffusiveFluxIntegral.C (hunk 3 above)
Comment (GiudGiud) [reply framework/src/postprocessors/SideDiffusiveFluxIntegral.C:89]: good catch. Should be d_grad_u(i)

File: framework/include/postprocessors/SideDiffusiveFluxIntegral.h [hunk 4]
```diff
@@ -36,5 +38,12 @@ class SideDiffusiveFluxIntegralTempl : public SideIntegralVariablePostprocessor
...
   MaterialPropertyName _diffusivity;
-  const GenericMaterialProperty<Real, is_ad> & _diffusion_coef;
+  const GenericMaterialProperty<T, is_ad> & _diffusion_coef;
+
+private:
+  /// Routine to get the diffusive flux with a Real diffusivity
+  RealVectorValue DiffusivityGradientProduct(RealVectorValue grad_u, Real diffusivity);
+
+  /// Routine to get the diffusive flux with a RealVectorValue diffusivity
+  RealVectorValue DiffusivityGradientProduct(RealVectorValue grad_u, RealVectorValue diffusivity);
```
Comment (lindsayad) [inline framework/include/postprocessors/SideDiffusiveFluxIntegral.h:48]:
```suggestion
  RealVectorValue diffusivityGradientProduct(const RealVectorValue & grad_u, Real diffusivity);

  /// Routine to get the diffusive flux with a RealVectorValue diffusivity
  RealVectorValue diffusivityGradientProduct(const RealVectorValue & grad_u, const RealVectorValue & diffusivity);
```

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/src/postprocessors/SideDiffusiveFluxIntegral.C [hunk 5]
```diff
@@ -60,11 +62,35 @@ SideDiffusiveFluxIntegralTempl<is_ad>::computeQpIntegral()
     const auto & grad_u = MetaPhysicL::raw_value(_fv_variable->adGradSln(*fi));
 
     // FIXME Get the diffusion coefficient on the face, see #16809
-    return -MetaPhysicL::raw_value(_diffusion_coef[_qp]) * grad_u * _normals[_qp];
+    return -DiffusivityGradientProduct(grad_u, MetaPhysicL::raw_value(_diffusion_coef[_qp])) *
```
Comment (lindsayad) [inline framework/src/postprocessors/SideDiffusiveFluxIntegral.C:65]:
```suggestion
    return -diffusivityGradientProduct(grad_u, MetaPhysicL::raw_value(_diffusion_coef[_qp])) *
```

File: framework/src/postprocessors/SideDiffusiveFluxIntegral.C [hunk 6]
```diff
@@ -60,11 +62,35 @@ SideDiffusiveFluxIntegralTempl<is_ad>::computeQpIntegral()
...
+           _normals[_qp];
   }
   else
-    return -MetaPhysicL::raw_value(_diffusion_coef[_qp]) * _grad_u[_qp] * _normals[_qp];
+    return -DiffusivityGradientProduct(_grad_u[_qp], MetaPhysicL::raw_value(_diffusion_coef[_qp])) *
```
Comment (lindsayad) [inline framework/src/postprocessors/SideDiffusiveFluxIntegral.C:69]:
```suggestion
    return -diffusivityGradientProduct(_grad_u[_qp], MetaPhysicL::raw_value(_diffusion_coef[_qp])) *
```

File: framework/src/postprocessors/SideDiffusiveFluxIntegral.C [hunk 7]
```diff
@@ -60,11 +62,35 @@ SideDiffusiveFluxIntegralTempl<is_ad>::computeQpIntegral()
...
+SideDiffusiveFluxIntegralTempl<is_ad, T>::DiffusivityGradientProduct(RealVectorValue grad_u,
+                                                                     Real diffusivity)
+{
+  return grad_u * diffusivity;
+}
+
+template <bool is_ad, typename T>
+RealVectorValue
+SideDiffusiveFluxIntegralTempl<is_ad, T>::DiffusivityGradientProduct(RealVectorValue grad_u,
+                                                                     RealVectorValue diffusivity)
```
Comment (lindsayad) [inline framework/src/postprocessors/SideDiffusiveFluxIntegral.C:84]:
```suggestion
SideDiffusiveFluxIntegralTempl<is_ad, T>::diffusivityGradientProduct(const RealVectorValue & grad_u,
                                                                     const Real diffusivity)
{
  return grad_u * diffusivity;
}

template <bool is_ad, typename T>
RealVectorValue
SideDiffusiveFluxIntegralTempl<is_ad, T>::diffusivityGradientProduct(const RealVectorValue & grad_u,
                                                                     const RealVectorValue & diffusivity)
```

File: test/tests/postprocessors/side_diffusive_flux_integral/tests [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+      input = 'side_diffusive_flux_integral.i'
+      exodiff = 'side_diffusive_flux_integral_out.e'
+      detail = 'with a scalar diffusivity,'
+    []
+    [vector_diffusivity]
+      type = 'Exodiff'
+      input = 'side_diffusive_flux_integral.i'
+      exodiff = 'side_diffusive_flux_integral_vector.e'
+      cli_args = "Outputs/file_base=side_diffusive_flux_integral_vector Postprocessors/avg_flux_right/type=SideVectorDiffusivityFluxIntegral Postprocessors/avg_flux_right/diffusivity=diffusivity_vec Postprocessors/inactive=''"
+      detail = 'and with an anisotropic vector diffusivity.'
```
Comment (lindsayad) [inline test/tests/postprocessors/side_diffusive_flux_integral/tests:17]: with the way things read in the actual sqa document produced, I would just do
```suggestion
      detail = 'with an anisotropic vector diffusivity.'
```

File: test/tests/postprocessors/side_diffusive_flux_integral/tests [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+      ad_indexing_type = 'global'
+      detail = 'with a scalar diffusivity,'
+    []
+    [vector_diffusivity]
+      type = 'Exodiff'
+      input = 'side_diffusive_flux_integral_fv.i'
+      exodiff = 'side_diffusive_flux_integral_fv_vector.e'
+      cli_args = "Outputs/file_base=side_diffusive_flux_integral_fv_vector Postprocessors/avg_flux_right/type=SideVectorDiffusivityFluxIntegral Postprocessors/avg_flux_right/diffusivity=diffusivity_vec Postprocessors/inactive=''"
+      ad_indexing_type = 'global'
+      detail = 'and with a vector diffusivity.'
```
Comment (lindsayad) [inline test/tests/postprocessors/side_diffusive_flux_integral/tests:37]:
```suggestion
      detail = 'with a vector diffusivity.'
```

**Review by lindsayad — CHANGES_REQUESTED** [review]

**Review by lindsayad — APPROVED** [review]

**Pull request merged**